test/test_selectors.py

```diff
@@ -0,0 +1,672 @@
+from __future__ import with_statement
+import os
+import psutil
+import signal
+import socket
+import sys
+import time
+import threading
+
+try:  # Python 2.6 unittest module doesn't have skip decorators.
+    from unittest import skip, skipIf, skipUnless
+    import unittest
+except ImportError:
+    from unittest2 import skip, skipIf, skipUnless
+    import unittest2 as unittest
+
+try:  # Python 2.x doesn't define time.perf_counter.
+    from time import perf_counter as get_time
+except ImportError:
+    from time import time as get_time
+
+try:  # Python 2.6 doesn't have the resource module.
+    import resource
+except ImportError:
+    resource = None
+
+try:  # Windows doesn't support socketpair on Python 3.5<
+    from socket import socketpair
+except ImportError:
+    from .socketpair_helper import socketpair
+
+from urllib3.util import selectors
+
+LONG_SELECT = 0.2
+SHORT_SELECT = 0.01
+
+# Tolerance values for CI timer/speed fluctuations.
+LOWER_TOLERANCE = 0.05
+UPPER_TOLERANCE = 0.05
+
+# Travis CI detection, sometimes Travis is extremely slow.
+if "TRAVIS" in os.environ:
+    UPPER_TOLERANCE = 0.55
+
+
+class AlarmThread(threading.Thread):
+    def __init__(self, timeout):
+        super(AlarmThread, self).__init__(group=None)
+        self.setDaemon(True)
+        self.timeout = timeout
+        self.canceled = False
+
+    def cancel(self):
+        self.canceled = True
+
+    def run(self):
+        time.sleep(self.timeout)
+        if not self.canceled:
+            os.kill(os.getpid(), signal.SIGALRM)
+
+
+class AlarmMixin(object):
+    def set_alarm(self, timeout):
+        if not hasattr(signal, "SIGALRM"):
+            self.skipTest("Platform doesn't have signal.SIGALRM")
+        if not hasattr(self, "alarm_thread"):
+            self.alarm_thread = None
```

Comment:
So this should be avoidable by having a class property called `alarm_thread` that is initialized to `None`. Any writes to `self.alarm_thread` won't overwrite the class variable, they'll just give a more specific value for that instance.